### PR TITLE
Generate Edit Page for profile

### PR DESCRIPTION
**Pull request by SharafiddinMukhtorov**

### CODE REVIEW

**Comment by Mirazyzz** [comment]
Please, attach a screenshot of a visual changes in your PR. You upload a screenshot as a comment to PR

image

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile.Android/Resources/drawable/back_icon.png
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile.Android/Resources/drawable/back_icon.png:1]: Red color should be used for danger actions, such as **Delete** or **Reject**, back button shouldn't be dangerous to click.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile.Android/Resources/values/styles.xml [hunk 1]
```diff
@@ -1,18 +1,23 @@
...
+	<!-- Base application theme -->
+	<style name="MainTheme.Base" parent="Theme.AppCompat.Light.DarkActionBar">
+		<!-- Add any other global styles here -->
+	</style>
+
+	<!-- Custom style for DatePickerDialog -->
+	<style name="CustomDatePickerDialog" parent="ThemeOverlay.MaterialComponents.MaterialAlertDialog">
+		<item name="colorAccent">#1C375C</item>
+		<item name="android:textColorPrimary">#1C375C</item>
+	</style>
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile.Android/Resources/values/styles.xml:21]: Have you tested the application? Are you sure this won't break any other places?

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/ApiClient.cs [hunk 2]
```diff
@@ -79,5 +79,41 @@ public async Task<HttpResponseMessage> PostAsync(string resource, string body)
                 throw;
             }
         }
+        public async Task<HttpResponseMessage> PutAsync(string resource, string body)
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/ApiClient.cs:82]: Separate with an empty line. Rename method to **UpdateAsync**, when giving a name for the method we should be platform, architecture agnostic, _Put_ is convention for REST API, it should affect us when writing code as a client, we only use it this name when provided by external libraries

Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/ApiClient.cs:82]: Separate with an empty line

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/ApiClient.cs [hunk 3]
```diff
@@ -79,5 +79,41 @@ public async Task<HttpResponseMessage> PostAsync(string resource, string body)
...
+        {
+            try
+            {
+                string token = await LocalStorage.GetAsync<string>(LocalStorageKey.Token);
+
+                if (string.IsNullOrEmpty(token))
+                {
+                    throw new InvalidTokenException("Token is empty.");
+                }
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/ApiClient.cs:91]: It looks like these lines and line **94** are good candidate for a separate method.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/ApiClient.cs [hunk 4]
```diff
@@ -79,5 +79,41 @@ public async Task<HttpResponseMessage> PostAsync(string resource, string body)
...
+                var request = new HttpRequestMessage(HttpMethod.Put, $"{BaseUrl}/{resource}");
+                request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", token);
+                request.Content = new StringContent(body, Encoding.UTF8, "application/json");
+
+                var response = await _client.SendAsync(request).ConfigureAwait(false);
+
+                if (!response.IsSuccessStatusCode)
+                {
+                    throw new HttpRequestException($"Failed to update data at {resource}. Status code: {response.StatusCode}");
+                }
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/ApiClient.cs:102]: maybe just **response.EnsureStatusCode()**?

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/ApiClient.cs [hunk 5]
```diff
@@ -79,5 +79,41 @@ public async Task<HttpResponseMessage> PostAsync(string resource, string body)
                 throw;
             }
         }
+        public async Task<HttpResponseMessage> PutAsync(string resource, string body)
+        {
+            try
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/ApiClient.cs:84]: No need for **try-catch** we won't do anything with an error in this method, all callers should take care of an exception themselves

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/ApiClient.cs (hunk 2 above)
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Services/ApiClient.cs:82]: We can directly deserialize the value returned from the API in the method and return a Generic parameter here. And accept a generic parameter for body and perform serialization and deserialization right here. so that the client would not bother about how our internal logic works.

`public async Task<TResult> PutAsync<TResult, TBody>(string resource, TBody body)`

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Stores/Account/AccountStore.cs [hunk 6]
```diff
@@ -110,5 +110,26 @@ private async Task<AccountDto> FetchDriverDataAsync(int accountId)
...
+            }
+
+            var json = JsonConvert.SerializeObject(account);
+            var response = await _client.PutAsync($"accounts/{account.Id}", json);
+
+            response.EnsureSuccessStatusCode();
+
+            var responseBody = await response.Content.ReadAsStringAsync();
+
+            var updatedAccount = JsonConvert.DeserializeObject<AccountDto>(responseBody);
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Stores/Account/AccountStore.cs:128]: if API client will work with generic types we will be able to replace all these lines with this.

`var response = await _client.PutAsync<AccountDto, AccountDto>($"accounts/{account.Id}", account);`

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Stores/Account/MockAccountStore.cs [hunk 7]
```diff
@@ -6,11 +6,33 @@ namespace CheckDrive.Mobile.Stores.Account
...
+            {
+                FirstName = "John",
+                LastName = "Doe",
+                Email = "john.doe@example.com",
+                Passport = "1234567891987654",
+                PhoneNumber = "+123456789",
+                Login = "johndoe",
+                Address = "123 Main St, Cityville",
+                Birthdate = new System.DateTime(1990, 1, 1),
+            };
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Stores/Account/MockAccountStore.cs:23]: Why don't you use Bogus library? We already have a helper class **FakeDataGenerator** which generates some fake data. You add some additional methods there if you need, or you can reuse what's already there.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ProfileViewModel.cs [hunk 8]
```diff
@@ -71,6 +78,33 @@ private async Task OnEditProfileAsync()
         {
             await _navigationService.NavigateToAsync(NavigationPageType.EditProfile);
         }
+        private async Task SaveEditsAsync()
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ProfileViewModel.cs:81]: this method should be wrapped into `try-catch` block and handle any errors that occurred.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ProfileViewModel.cs [hunk 9]
```diff
@@ -71,6 +78,33 @@ private async Task OnEditProfileAsync()
...
+            {
+                Login = Login,
+                FirstName = FullName.Split(' ')[0],
+                LastName = FullName.Split(' ').Length > 1 ? FullName.Split(' ')[1] : "",
+                Passport = Passport,
+                PhoneNumber = PhoneNumber,
+                Email = Email,
+                Address = Address,
+                Birthdate = DateTime.Parse(Birthdate),
+            };
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ProfileViewModel.cs:93]: maybe let's make it a private method for example **GetAccount()** to make this method smaller and more readable.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ProfileViewModel.cs [hunk 10]
```diff
@@ -71,6 +78,33 @@ private async Task OnEditProfileAsync()
...
+
+            var success = await _accountService.UpdateAccountAsync(updatedAccount);
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ProfileViewModel.cs:95]: let's rename it to **result** or something else, because _success_ is not clear.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ProfileViewModel.cs [hunk 11]
```diff
@@ -71,6 +78,33 @@ private async Task OnEditProfileAsync()
...
+
+            if (success != null)
+            {
+                await _navigationService.NavigateToAsync(NavigationPageType.Profile);
+                await Application.Current.MainPage.DisplayAlert("Success", "Profil muvafaqiyatli tarzda o'zgartirildi.", "OK");
+            }
+            else
+            {
+                await _navigationService.NavigateToAsync(NavigationPageType.Profile);
+                await Application.Current.MainPage.DisplayAlert("Error", "Profil o'zgartirishda muammo aniqlandi. Iltimos qaytadan urinib ko'ring.", "OK");
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ProfileViewModel.cs:105]: this logic should be in `catch`

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/EditProfile.xaml
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/EditProfile.xaml:1]: please attach screenshot of this page.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/EditProfile.xaml.cs [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+using Xamarin.Forms;
+using Xamarin.Forms.Xaml;
+
+namespace CheckDrive.Mobile.Views
+{
+    [XamlCompilation(XamlCompilationOptions.Compile)]
+    public partial class EditProfile : ContentPage
+    {
+        private readonly ProfileViewModel _viewModel;
+        public EditProfile()
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/EditProfile.xaml.cs:17]: separate with an empty line

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/EditProfile.xaml.cs [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+        {
+            InitializeComponent();
+
+            _viewModel = new ProfileViewModel();
+
+            BindingContext = _viewModel;
+        }
+        protected override async void OnAppearing()
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/EditProfile.xaml.cs:25]: separate with an empty line.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/EditProfile.xaml.cs [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+        {
+            if (BindingContext is ProfileViewModel vm)
+            {
+                await vm.LoadProfileDataAsync();
+            }
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/Views/EditProfile.xaml.cs:30]: we already have it in **_viewModel** no need for casting.

`await _viewModel.LoadProfileDataAsync();`

**Review by Mirazyzz — CHANGES_REQUESTED** [review]
Please, provide necessary changes, and upload screenshots of visual changes you have made.

**Comment by SharafiddinMukhtorov** [comment]
Screenshot 2024-10-01 131808

**Comment by SharafiddinMukhtorov** [comment]
Edit Page Calendar Design

**Comment by SharafiddinMukhtorov** [comment]
Screenshot 2024-10-01 163130
Edit Page Design

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile.Android/MainActivity.cs [hunk 15]
```diff
@@ -2,27 +2,24 @@
...
+    {
+        Xamarin.Essentials.Platform.OnRequestPermissionsResult(requestCode, permissions, grantResults);
 
-            base.OnRequestPermissionsResult(requestCode, permissions, grantResults);
-        }
+        base.OnRequestPermissionsResult(requestCode, permissions, grantResults);
     }
-}
\ No newline at end of file
+}
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile.Android/MainActivity.cs:25]: Revert back formatting changes.

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile.Android/MainActivity.cs [hunk 16]
```diff
@@ -2,27 +2,24 @@
 using Android.Content.PM;
 using Android.OS;
 using Android.Runtime;
-using CheckDrive.Mobile.ViewModels;
+using CheckDrive.Mobile;
 
-namespace CheckDrive.Mobile.Droid
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile.Android/MainActivity.cs:7]: why did you delete namespace?

File: CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ProfileViewModel.cs [hunk 17]
```diff
@@ -72,6 +79,29 @@ private async Task OnEditProfileAsync()
...
+            catch (Exception ex)
+            {
+                await _navigationService.NavigateToAsync(NavigationPageType.Profile);
+                await Application.Current.MainPage.DisplayAlert("Error", $"Profil o'zgartirishda muammo aniqlandi. Iltimos qaytadan urinib ko'ring. Xato: {ex.Message}", "OK");
+            }
+        }
+
         private async Task OnLogoutAsync()
         {
             var confirmed = await Application.Current.MainPage.DisplayAlert("Logout", "Are you sure you want to log out?", "Yes", "No");
```
Comment (Mirazyzz) [inline CheckDrive.Mobile/CheckDrive.Mobile/CheckDrive.Mobile/ViewModels/ProfileViewModel.cs:107]: text should be translated

**Review by Mirazyzz — CHANGES_REQUESTED** [review]
Please, check each comment and make necessary changes.

**Review by Mirazyzz — CHANGES_REQUESTED** [review]
Please, provide necessary changes

**Review by Mirazyzz — APPROVED** [review]
